openfl/utilities/fed_timer.py

```diff
@@ -0,0 +1,126 @@
+            self._duration = round(time.perf_counter() - start, 3)
+
+    def result(self):
+        return self._result, self._duration
+
+class PrepareTask():
+
+    def __init__(self, target_fn, timeout, args, kwargs) -> None:
+        self._target_fn = target_fn
+        self._decorated_fn_name = target_fn.__name__
+        self._is_coroutine = asyncio.iscoroutinefunction(target_fn)
+        self._max_timeout = timeout
+        self._args = args
+        self._kwargs = kwargs
+        self._result = None
+        self._duration = None
+
+    async def async_execute(self):
+
+        start = time.perf_counter()
+
+        task = asyncio.create_task(
+            name=self._decorated_fn_name,
+            coro=self._target_fn(*self._args, **self._kwargs)
+        )
+
+        try:
+            await asyncio.wait_for(task, timeout=self._max_timeout)
+            self._result = task.result()
+            self._duration = time.perf_counter() - start
+        except asyncio.TimeoutError as exc:
+            raise asyncio.TimeoutError(f"Exception Timeout Terminating Method: {self._decorated_fn_name}, Co-routine: {self._is_coroutine}")
+        except Exception as exc:
+            raise Exception(f"Generic Exception: {self._decorated_fn_name}, Co-routine: {self._is_coroutine}")
+
+    def sync_execute(self):
+
+        task = CustomThread(target=self._target_fn,
+                            name=self._decorated_fn_name,
+                            args=self._args,
+                            kwargs=self._kwargs)
+        task.start()
+        task.join(self._max_timeout)
+
+        self._result, self._duration = task.result()
+
+        if task.is_alive():
+            logger.info(f"Logger Terminating: {task.name}")
+            # print(traceback.format_exc())
+            print(f"Timeout Terminating: {task.name}", file=sys.stderr)
+            raise TimeoutError(f"Exception Timeout Terminating Method: {task.name}, Co-routine: {self._is_coroutine}")
+
+    def result(self):
+        if self._duration is None:
+            self._duration = self._max_timeout
+        return self._result, self._duration
+
+def parameterize(dec):
+    def layer(*args, **kwargs):
+        def repl(f):
+            return dec(f, *args, **kwargs)
+        return repl
+    return layer        
+
+@parameterize
+def fedtiming(func, timeout):
+
+    def init_task(*args, **kwargs):
+        return PrepareTask(target_fn=func, timeout=timeout, args=args, kwargs=kwargs)
+
+    def exception_handler(exc):
+        template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+        message = template.format(type(exc).__name__, exc.args)
+        # print(message)
+        logger.exception(traceback.format_exc())
+        time.sleep(2)
+        os._exit(status=os.EX_TEMPFAIL)
+
+    async def async_wrapper(*args, **kwargs):
+        task = init_task(*args, **kwargs)
+        logger.info(f"Started Async: ({func.__name__})")
```

Comment:
Although, I didn't face any issue yet. Skimming through the code, I am suspecting the `func.__name__` might throw an exception if you happen to decorate `@fedtiming(timeout)` to a method which is already decorated with `@staticmethod`. Do let me know, if you encounter this. It's a minor change, incase.